py/selenium/webdriver/firefox/webdriver.py

```diff
@@ -47,35 +48,115 @@ class WebDriver(RemoteWebDriver):
+            or a string.  If undefined, a fresh profile will be created
+            in a temporary location on the system.
+        :param firefox_binary: Instance of ``FirefoxBinary`` or full
+            path to the Firefox binary.  If undefined, the system default
+            Firefox installation will  be used.
+        :param timeout: Time to wait for Firefox to launch when using
+            the extension connection.
+        :param capabilities: Dictionary of desired capabilities.
+        :param proxy: The proxy settings to us when communicating with
+            Firefox via the extension connection.
+        :param executable_path: Full path to override which geckodriver
+            binary to use for Firefox 47.0.1 and greater, which
+            defaults to picking up the binary from the system path.
+        :param firefox_options: Instance of ``options.Options``.
+        :param log_path: Where to log information from the driver.
+
+        """
+        self.binary = None
+        self.profile = None
+
+        if capabilities is None:
+            capabilities = DesiredCapabilities.FIREFOX.copy()
+        if firefox_options is None:
+            firefox_options = Options()
+
+        if capabilities.get("binary"):
+            self.binary = capabilities["binary"]
+
+        # firefox_options overrides capabilities
+        if firefox_options is not None:
+            if firefox_options.binary is not None:
+                self.binary = firefox_options.binary
+            if firefox_options.profile is not None:
+                self.profile = firefox_options.profile
+
+        # firefox_binary and firefox_profile
+        # override firefox_options
+        if firefox_binary is not None:
+            if isinstance(firefox_binary, types.StringTypes):
+                firefox_binary = FirefoxBinary(firefox_binary)
+            self.binary = firefox_binary
+            firefox_options.binary = firefox_binary
+        if firefox_profile is not None:
+            if isinstance(firefox_profile, types.StringTypes):
+                firefox_profile = FirefoxProfile(firefox_profile)
+            self.profile = firefox_profile
+            firefox_options.profile = firefox_profile
+
+        # W3C remote
+        # TODO(ato): Perform conformance negotiation
         if capabilities.get("marionette"):
-            self.service = Service(executable_path, firefox_binary=self.options.binary_location, log_path=log_path)
+            self.service = Service(executable_path, log_path=log_path)
             self.service.start()
 
+            capabilities.update(firefox_options.to_capabilities())
+
             executor = FirefoxRemoteConnection(
                 remote_server_addr=self.service.service_url)
             RemoteWebDriver.__init__(
                 self,
                 command_executor=executor,
                 desired_capabilities=capabilities,
                 keep_alive=True)
+
+            self._is_remote = True
```

Comment:
This needs to be false.

`self._is_remote` is used to know when we need to ship a file over to a selenium server on a different machine.